File: android/app/src/main/java/be/sportgreenmoove/app/services/ActiveRideStartup.kt
package be.sportgreenmoove.app.services

import be.sportgreenmoove.app.data.AppRole
import be.sportgreenmoove.app.data.LiveRideSnapshot
import be.sportgreenmoove.app.data.RideCompletionSummary

data class ActiveRideStartResult(
    val ride: LiveRideSnapshot,
    val notice: String,
)

suspend fun AndroidProviderSet.startTrackedRide(
    tripId: String,
    role: AppRole,
    bookingIds: List<String>,
): ActiveRideStartResult {
    val ride = firebase.startRide(tripId, bookingIds)
    return startTrackingForRide(ride, role)
}

suspend fun AndroidProviderSet.startAccessibleRideTracking(
    role: AppRole,
): ActiveRideStartResult {
    val ride = firebase.getActiveRide()
        ?: throw ProviderConfigurationException("Aucune course active accessible pour démarrer le suivi.")
    return startTrackingForRide(ride, role)
}

private suspend fun AndroidProviderSet.startTrackingForRide(
    ride: LiveRideSnapshot,
    role: AppRole,
): ActiveRideStartResult {
    val radarStarted = if (radar.isConfigured) {
        runCatching {
            radar.startTripTracking(ride.rideSessionId, role)
        }.isSuccess
    } else {
        false
    }

    firebase.writeNativeLocationFallback(ride.rideSessionId, role)

    val notice = if (radarStarted) {
        "Suivi Radar et secours GPS Firebase activés."
    } else {
        "Secours GPS Firebase activé."
    }
    return ActiveRideStartResult(ride = ride, notice = notice)
}

suspend fun AndroidProviderSet.endTrackedRide(
    rideSessionId: String,
    distanceMeters: Int,
    passengersSharing: Int,
): RideCompletionSummary {
    val completion = firebase.endRide(
        rideSessionId = rideSessionId,
        distanceMeters = distanceMeters,
        passengersSharing = passengersSharing,
    )
    firebase.stopNativeLocationFallback(rideSessionId)
    if (radar.isConfigured) {
        runCatching { radar.stopTripTracking(rideSessionId) }
    }
    return completion
}

File: ios/Sources/SportsGreenMooveApp/ActiveRideStartup.swift
import Foundation

struct ActiveRideStartResult: Sendable {
    let ride: LiveRideSnapshot
    let notice: String
}

func startTrackedRide(
    firebase: FirebaseGateway,
    radar: RadarTrackingGateway,
    tripId: String,
    role: AppRole,
    bookingIds: [String]
) async throws -> ActiveRideStartResult {
    let ride = try await firebase.startRide(tripId: tripId, bookingIds: bookingIds)
    return try await startTrackingForRide(ride, firebase: firebase, radar: radar, role: role)
}

func startAccessibleRideTracking(
    firebase: FirebaseGateway,
    radar: RadarTrackingGateway,
    role: AppRole
) async throws -> ActiveRideStartResult {
    guard let ride = try await firebase.getActiveRide() else {
        throw ProviderConfigurationError(message: "Aucune course active accessible pour démarrer le suivi.")
    }
    return try await startTrackingForRide(ride, firebase: firebase, radar: radar, role: role)
}

private func startTrackingForRide(
    _ ride: LiveRideSnapshot,
    firebase: FirebaseGateway,
    radar: RadarTrackingGateway,
    role: AppRole
) async throws -> ActiveRideStartResult {
    let radarStarted: Bool
    if radar.isConfigured {
        radarStarted = (try? await radar.startTripTracking(rideSessionId: ride.rideSessionId, role: role)) != nil
    } else {
        radarStarted = false
    }

    try await firebase.writeNativeLocationFallback(rideSessionId: ride.rideSessionId, role: role)

    let notice = radarStarted
        ? "Suivi Radar et secours GPS Firebase activés."
        : "Secours GPS Firebase activé."
    return ActiveRideStartResult(ride: ride, notice: notice)
}

func endTrackedRide(
    firebase: FirebaseGateway,
    radar: RadarTrackingGateway,
    rideSessionId: String,
    distanceMeters: Int,
    passengersSharing: Int
) async throws -> RideCompletionSummary {
    let completion = try await firebase.endRide(
        rideSessionId: rideSessionId,
        distanceMeters: distanceMeters,
        passengersSharing: passengersSharing
    )
    firebase.stopNativeLocationFallback(rideSessionId: rideSessionId)
    if radar.isConfigured {
        try? await radar.stopTripTracking(rideSessionId: rideSessionId)
    }
    return completion
}
